lib/puppet/indirector/rest.rb

```diff
@@ -249,7 +250,21 @@ def save(request)
   # to request.do_request from here, thus if we change what we pass or how we
   # get it, we only need to change it here.
   def do_request(request)
-    request.do_request(self.class.srv_service, self.class.server, self.class.port) { |req| yield(req) }
+    response = request.do_request(self.class.srv_service, self.class.server, self.class.port) { |req| yield(req) }
+
+    handle_response(request, response) if response
+
+    response
+  end
+
+  def handle_response(request, response)
+    server_version = response[Puppet::Network::HTTP::HEADER_PUPPET_VERSION]
+    if server_version &&
+       SemanticPuppet::Version.parse(server_version).major < 5 &&
+       Puppet[:preferred_serialization_format] != 'pson'
+      Puppet.warning("Downgrading to PSON for future requests")
+      Puppet[:preferred_serialization_format] = 'pson'
```

Comment:
Is there any chance we might inadvertently overwrite a user's setting here when they don't want us to? I.e., if they've specified `:preferred_serialization_format = msgpack`.

Even if that were the case though, I'm not sure what we'd do about it.. unless the response contains the supported formats, and then we could check if the current supported format is not among them?